fastcrypto-tbls/src/nidkg.rs

```diff
@@ -0,0 +1,493 @@
+    pub fn new<R: AllowedRng>(
+        ecies_sk: ecies::PrivateKey<G>,
+        nodes: Vec<Node<G>>,
+        t: u32, // The number of shares that are needed to reconstruct the full signature.
+        random_oracle: RandomOracle,
+        rng: &mut R,
+    ) -> Result<Self, FastCryptoError> {
+        let ecies_pk = ecies::PublicKey::<G>::from_private_key(&ecies_sk);
+        let my_id = nodes
+            .iter()
+            .find(|n| n.pk == ecies_pk)
+            .ok_or(FastCryptoError::InvalidInput)?
+            .id;
+        let nodes = Nodes::new(nodes)?;
+        let n = nodes.n();
+        if t >= n {
+            return Err(FastCryptoError::InvalidInput);
+        }
+        let vss_sk = PrivatePoly::<G>::rand(t - 1, rng);
+
+        // Precompute the dual code coefficients.
+        let ids_as_scalars = (1..=n)
+            .into_iter()
+            .map(|i| (i, G::ScalarType::from(i as u64)))
+            .collect::<HashMap<_, _>>();
+        let precomputed_dual_code_coefficients = (1..=n)
+            .into_iter()
+            .map(|i| {
+                (1..=n)
+                    .into_iter()
+                    .filter(|j| i != *j)
+                    .map(|j| ids_as_scalars[&i] - ids_as_scalars[&j])
+                    .fold(G::ScalarType::generator(), |acc, x| acc * x)
+                    .inverse()
+                    .expect("non zero")
+            })
+            .collect();
+
+        Ok(Self {
+            id: my_id,
+            nodes,
+            t,
+            random_oracle,
+            ecies_sk,
+            vss_sk,
+            precomputed_dual_code_coefficients,
+        })
+    }
+
+    /// 4. Create the message to be sent to all parties.
+    pub fn create_message<R: AllowedRng>(&self, rng: &mut R) -> Message<G> {
+        // TODO: Can this be done faster?
+        let partial_pks = self
+            .nodes
+            .share_ids_iter()
+            .map(|i| G::generator() * self.vss_sk.eval(i).value)
+            .collect();
+        let mut interim_values: Vec<[(G::ScalarType, G); NUM_OF_ENCRYPTIONS_PER_SHARE]> =
+            Vec::new();
+        let pairs = self
+            .nodes
+            .share_ids_iter()
+            .map(|share_id| {
+                let mut values = Vec::new();
+                let node = self
+                    .nodes
+                    .share_id_to_node(&share_id)
+                    .expect("using valid share id");
+                let encryptions = (0..NUM_OF_ENCRYPTIONS_PER_SHARE)
+                    .into_iter()
+                    .map(|_| {
+                        let r = G::ScalarType::rand(rng);
+                        let msg = bcs::to_bytes(&r).expect("serialization should work");
+                        let r_g = G::generator() * r;
+                        let r_x_g = *node.pk.as_element() * r;
+                        // Save also the points instead of recomputing them later.
+                        values.push((r, r_x_g));
```

Comment:
I think the values vector is not needed - the values can be pushed to interim_values directly.

Reply:
note that interim_values is a vector of fixed size arrays